nolearn/lasagne/base.py

```diff
@@ -46,17 +54,24 @@ def __call__(self, X, y=None):
         return self
 
     def __iter__(self):
-        n_samples = self.X.shape[0]
         bs = self.batch_size
-        for i in range((n_samples + bs - 1) // bs):
+        for i in range((self.n_samples + bs - 1) // bs):
             sl = slice(i * bs, (i + 1) * bs)
-            Xb = self.X[sl]
+            Xb = _sldict(self.X, sl)
             if self.y is not None:
                 yb = self.y[sl]
             else:
                 yb = None
             yield self.transform(Xb, yb)
 
+    @property
+    def n_samples(self):
+        X = self.X
+        if isinstance(X, dict):
+            return len(list(X.values())[0])
```

Comment:
On the `__call__` for the base iterator perhaps we should assert that all X's have the same length?